rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,518 @@
+                                                              │KubeAgent│                                        │Teleport│          │Kubernetes│                    
+                                                              └─────────┘                                        └────────┘          └──────────┘                    
+```
+
+### Backend storage
+
+Kubernetes Secret storage will be a transparent backend storage that can be plugged into [`ProcessStorage`](https://github.com/gravitational/teleport/blob/1aa38f4bc56997ba13b26a1ef1b4da7a3a078930/lib/auth/state.go#L35) structure. Currently, `ProcessStorage` is responsible for handling the Identity and State storage. It expects a [`backend.Backend`](https://github.com/gravitational/teleport/blob/cc27d91a4585b0d744a7bec110260c335b0007fd/lib/backend/backend.go#L42), but it only requires a subset of it. The idea is to define a new interface `IdentityBackend` with the required methods and use it.
+
+```go
+// ProcessStorage is a backend for local process state,
+// it helps to manage rotation for certificate authorities
+// and keeps local process credentials - x509 and SSH certs and keys.
+type ProcessStorage struct {
+	IdentityBackend
+}
+
+// IdentityBackend implements abstraction over local or remote storage backend methods
+// required for Identity/State storage.
+// As in backend.Backend, Item keys are assumed to be valid UTF8, which may be enforced by the
+// various Backend implementations.
+type IdentityBackend interface {
+	// Create creates item if it does not exist
+	Create(ctx context.Context, i backend.Item) (*backend.Lease, error)
+	// Put puts value into backend (creates if it does not
+	// exists, updates it otherwise)
+	Put(ctx context.Context, i backend.Item) (*backend.Lease, error)
+	// Get returns a single item or not found error
+	Get(ctx context.Context, key []byte) (*backend.Item, error)
+	// Close closes backend and all associated resources
+	Close() error
+}
+```
+
+During the startup procedure, the agent identifies that it is running inside Kubernetes. The identification can be achieved by checking the presence of the service account mount path `/var/run/secrets/kubernetes.io`. Although Kubernetes has an option that can disable this mount path, `automountServiceAccountToken: false`, this option is always `true` in our helm chart since we require it for handling the secrets.
+
+If the agent detects that it is running in Kubernetes, it instantiates a backend for Kubernetes Secret. This backend creates a client with configuration provided by `restclient.InClusterConfig()`, which uses the service account token mounted in the pod. With this, the agent can operate the secret by creating, updating, and reading the secret data. To prevent multiple agents from racing each other when writing in the secret, the Kubernetes Storage engine might use the resource lock feature from Kubernetes (`resourceVersion`) to implement optimistic locking.
```

Comment:
Yes, and it can happen if we are running Deployment and the operator scales it without using helm. In that case, multiple replicas will use exactly the same secret because the `TELEPORT_REPLICA_NAME` is constant.